#### Feature Selection With Filtering Method

### Constant , Quasi constant and Duplicate Feature Removal

Machine learning and deep learning is learning from data which consist of diffrent types of features.

lets say if we have huge datasets it could have huge featurespace as well.if we have huge feature space then we want some types of algorithms which can select the highest performing features for apping machine learning algorithms otherwise u will endup your machine  with overfitting which will deterioate the performace of your model.
Training time and Performace of our machine learning depends havily on the features in the datasets.
IF we having the large datasets then it will definately take long traning time.
Its is garanteed that huge feature space will result the better performace.
So idealy we have to select the those feature which are actually help in machine learning model learn something usefull.

Unnecessary and redudant features not only slow down the traing time pf our algorithms but they are also affect the overall performace of the algorithm.

The process of selecting the most suitable feature for traing the machine learning model is called a feature selection.

In that we are going to learn about the feature selection with the filtering method.In that we will be do the constant,quasi constant and duplicate feature removal.

There is several advantages of performing the feature selection before taring our machine learning model.

1.Models with the less numbers of featutres have higher explainability.

2.Its very easier to implement the machine lerning models with the reduced features.

3.Fewer feature leads to enhance the generalization which in turns reduces overfitting.

4.Feature selection reduces data redudancy.

5.Training time for a model with fewer features is significantly lower.

6.Models with the fewer features are less prone to errors.

#### What is Filter Method ?

Filter method is belonging to the feature selection method that select features independently of machine learning models.
It doesn't involve any machine learning alorithm while selecting a features.
This is one of the biggest advantage of this method since is doesnt inovolve any machine learning algorithm that's it is a pretty fast.
Feature selected by using with the filter method can be use as an input to any machine learning model.
And another advantages of the fiter method is they are very fast.
Filter method is a generally first step in feature selection pipeline so this kind of best screening method to get subset of the features.

There are two kind of feature seection method in filtering method those are the 


1.Univariate

2.Multi-Variate

In the univariate we are selecting the feature without taking consideration of other features.Thats it kind of independed.
lets consider we having the 10 features then we can select the some subset of the features without taking the information of nine features for particular feature.
In the univariant feature selection we can select the features based on the Fisher score,Mutual Information Gain etc.


The univariate filter methods are types of the method where individual features are ranked according to the specific criteria.The top N features are then selected.Diffrent type of the ranking criteria are used for the univariat filter methods.For an example:- Fisher Score,mutual information and variance of the feature.


Multi-Variate

In in multi-variate feature selection method we take information of feature to feature.Then we will try to find out the best feature out of those subset features and method from which we select known as "pearson correletion"

Multivariant filter methods are capable of removing redundant features from the data since they mutual relationship between the features into account.

### Univariate Filtering Methods

Constant Removal

Quasi Constant Removal

Duplicate Feature Removal

### Constant Feature Removal

In [3]:
#import the all liberary
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import VarianceThreshold

In [6]:
#load the datasets
#n_rows for the simplicity

In [7]:
data = pd.read_csv('santander-train.csv',nrows=2000)
#we are using the random forest classifier :- it does not waant the standard scale or standardize the our datasets.
#since,tree based model doesn't consider scales of various feature.
#means standardization and normalization not required in the tree based model.
data.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                        0                        0  ...   
1                        0                        0  ...   
2                        0                        0  ...   
3                        0                        0  ...   
4                        0                        0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [8]:
x = data.drop('TARGET',axis = 1)
y=data['TARGET']
x.shape, y.shape

((2000, 370), (2000,))

AFter we have to apply the to split the data into the training and testing 
we have to apply the some operation on the test dataset part to avoid the overfitting.

In [9]:
x_train,x_test,y_train,y_test = train_test_split(x,y ,test_size=.2,train_size=.8,stratify=y,random_state =0)

#### Constant Features Removal

In [ ]:
#here we have to first use the variance thresold ,fit and then use transform.
#thresold variance is always work on the training datasets.

In [10]:
constant_filter = VarianceThreshold(threshold=0)
constant_filter.fit(x_train)

VarianceThreshold(threshold=0)

In [12]:
#in this way we can get the total number of the features after removing the constant features.
constant_filter.get_support().sum()
#means it return non_constant features.

261

In [ ]:
#if invert the above code we will get the constant features.

In [13]:
constant_list = [not temp for temp in constant_filter.get_support()]
constant_list
#it return the true for the those value who are the constant.

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 Fals

In [14]:
#if we want to see those varaible 
x.columns[constant_list]

Index(['ind_var2_0', 'ind_var2', 'ind_var13_medio_0', 'ind_var13_medio',
       'ind_var18_0', 'ind_var18', 'ind_var27_0', 'ind_var28_0', 'ind_var28',
       'ind_var27',
       ...
       'saldo_medio_var13_medio_hace3', 'saldo_medio_var13_medio_ult1',
       'saldo_medio_var13_medio_ult3', 'saldo_medio_var17_hace3',
       'saldo_medio_var29_hace2', 'saldo_medio_var29_hace3',
       'saldo_medio_var33_hace2', 'saldo_medio_var33_hace3',
       'saldo_medio_var33_ult1', 'saldo_medio_var33_ult3'],
      dtype='object', length=109)

In [ ]:
#now lets go ahead transform the x_train datasets into non-coonstant feature space 
#remove all the constant feature space.

In [15]:
#to avoid the overfitting we had used the the transform method #we appllied on traing dataset as well on the test to avoid the overfitting.
x_train_filter = constant_filter.transform(x_train)
x_test_filter = constant_filter.transform(x_test)
#so,lets check about the features avalibilities in the x_train_filter,y_tain_filter and compare with the x_train
x_train_filter.shape , x_test_filter.shape,x_train.shape

#now can see that there is a constant feature are removed from the x_train and x_test datasets by using the varaiancethresold and then fit and transform.
#whereas, the previous traing dataset had 370 features outof which the constant features are removed completely.

((1600, 261), (400, 261), (1600, 370))

### Quasi Constant Feature Removal

Quasi constant menas:-

The name suggest that the features that almost constant 
In another words, this feature have the same value for very large subset of the output and such are not really usefull for making predictions .This features create unnecessarily overhead on the machine learning model.

There is no rule as to waht we should be thresold we select for the quasi constant feature but in general and widely we use the 0.01 % of variance thresold is that means 1% of similarity .
that means if we select the 0.01 value of the variance that time we remove the all feature set they are 99% similar.

A feature content is always content a single value which is almost constant but there is very a small chance that feature is shifted vary somehow from the mean that means variance would be the very low that kind feature we say as quasi constant
and those features are somehow near to the constant.

In [ ]:
#so lets do something for the quasi constant feature removal 

In [16]:
quasi_constant_filter = VarianceThreshold(threshold=0.01)

In [17]:
quasi_constant_filter.fit(x_train_filter)

VarianceThreshold(threshold=0.01)

In [18]:
quasi_constant_filter.get_support().sum()

223

If we kindly observe that the we having the remaing after the removal of the quasi constant features that are 223 before that we had feature 261 to the x_train_filter i.e.261-223= 38 features here we removed those are the quasi constant.

In [19]:
x_train_quasi_filter = quasi_constant_filter.transform(x_train_filter)
x_test_quasi_filter = quasi_constant_filter.transform(x_test_filter)
x_train_quasi_filter.shape  , x_test_quasi_filter.shape

((1600, 223), (400, 223))

This is the way we had removed the constant and quasi constant features.

### Duplicate Features Removal

If the two features are exacly same that means they are duplicate.That means the duplicate feature does not introdue the any information but it increases the overhead of the machine learning model.In that way it can creates the complex model by increasing the training time  but overall performance would be reduce in case if our datasets having some duplicate feature.

Problem is that the sklern or python doesnt have any liberary to remove duplicate of the feature datasets.So for that we have to use some transpose technique,so lets see

Transpose- rows will get the place of the column and vice versa.

In [20]:
x_train_T = x_train_quasi_filter.T
x_test_T  = x_test_quasi_filter.T

In [21]:
#by using the transpose - pandas dataframe become array.
type(x_train_T)

numpy.ndarray

In [22]:
#so,we need to convert them into the dataframe format
x_train_T = pd.DataFrame(x_train_T)
x_test_T = pd.DataFrame(x_test_T)

In [23]:
#if we want to check about the rows and columns
x_train_T.shape , x_test_T.shape
#now can see that the row become columns and column become rows

((223, 1600), (223, 400))

In [24]:
#to chck how many features are duplicated in the dataframe for that we need to check by using the duplicated method.
x_train_T.duplicated().sum()

23

In [25]:
#this is the way we can get the duplicated feature if the true that means its duplicated if not then its not duplicated.
duplicated_feature = x_train_T.duplicated()
duplicated_feature

0      False
1      False
2      False
3      False
4      False
       ...  
218    False
219    False
220    False
221    False
222    False
Length: 223, dtype: bool

In [26]:
#if we want to get the info about the non-duplicated featues so for that here we are founding the all the true value are non-duplicated
#and duplicated are in the form of false.
features_to_keep = [not index for index in duplicated_feature]
features_to_keep

[True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 False,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 False,
 True,
 False,
 False,
 True,
 False,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,


In [27]:
x_train_unique = x_train_T[features_to_keep].T
x_test_unique  = x_test_T[features_to_keep].T

x_train_unique.shape,x_test_unique.shape

((1600, 200), (400, 200))

### Build ML model and compare the performance of the selected features.

In [58]:
def run_RandomForest(x_train,x_test,y_train,y_test):
    clf = RandomForestClassifier(n_estimators=100,random_state=0,n_jobs=-1)
    clf.fit(x_train,y_train)
    y_pred = clf.predict(x_test)
    print('Accuracy of RFC',accuracy_score(y_test,y_pred))
#we have to run the function:
run_RandomForest(x_train_unique,x_test_unique,y_train,y_test)

UsageError: Line magic function `%%time` not found.


In [57]:
#to know the training time we have to use magic time with the help of the time function.
%%time
run_RandomForest(x_train_unique,x_test_unique,y_train,y_test)

Accuracy of RFC 0.955
Wall time: 790 ms


In [ ]:
#if we real dataset then we have check the diffrence between them:-

In [69]:
%%time
run_RandomForest(x_train,x_test,y_train,y_test)

Accuracy of RFC 0.9575
Wall time: 846 ms


In [ ]:
#from we can assume that the feature selection method may not significantly inpact on the accuracy but it significantly imact on the
#training time.